### ISSUE TITLE
mass_loss.c

### ISSUE THREAD

**Issue by sabaronett** [sabaronett/reboundx#5]
@Nofe4108 
Thought we should create its own issue to track your progress/discussions on this front.

By the way, did you try out this REBOUNDx example and/or modify it or use it as a template/basis for your eventual mass loss effect?

https://reboundx.readthedocs.io/en/latest/effects.html#mass-modifications

**Comment by Nofe4108** [comment]
Sounds good. I did try out that example, and I've been using it as a template as I started making my effect. Problem is my own effect doesn't seem to be working. I made it as simple as possible, but when I run the basic simulation I have paired with it the effect doesn't seem to be doing anything at all.

**Comment by sabaronett** [comment]
@Nofe4108 , have you verified the original example (before you made any changes), how it works, and what to look for in the sim that uses it?

In the meantime, go ahead and,

1. Make a new branch for your effect on this repo like you did on your fork (e.g., `mass_loss`)
2. Commit and push the changes you've made thus far. (Let me know if you run into any issues with 1. or 2.)
3. If you're still stuck, make a comment on this issue with the following:
     - Cite your last commit like this https://github.com/sabaronett/reboundx/commit/59fc9550f01b2aa2b29787591c9414f834651be4 (edit my comment to see the example URL).
     - Detail what's wrong, or unexpected behavior, and any clues/thoughts on where to look next (if possible)
4. We'll invite Dan Tamayo to look at this issue and ask for any insight as to what to check/do next.

**Comment by sabaronett** [comment]
@Nofe4108 
So I played around with `problem.c` in the existing `example/modify_mass` and discovered that the type of integrator set seems to make or break the REBx effect (see https://github.com/sabaronett/reboundx/commit/b66ca7016df691443313b1a639a3bb21de957496)
https://github.com/sabaronett/reboundx/blob/b66ca7016df691443313b1a639a3bb21de957496/examples/modify_mass/problem.c#L24-L36

So far, it seems that the `modify_mass` effect won't work with `_MERCURIUS` or `_EOS` (their newest integrator). However, I've confirmed that (as you can try for yourself by running `./rebound` for several seconds or minutes, then running `python sim_reader.py`) **both** `_WHFAST` and `_IAS15` work, although they run _significantly_ slower. I suspect it has to do with which ones use _adaptive timesteps_ and which don't; maybe the adaptive timesteps assume no changes in mass to speed up the integrations, thus effectively ignoring the effect entirely.

I've created another issue (#9) for us to follow up on which integrators may or may not work with our final effects, but, in the meantime, let's stick with either of these two integrators for our tests. In fact, let me know if by switching integrators, your `mass_loss` code works now.

**Comment by Nofe4108** [comment]
Hey, sorry I haven't been able to work for a few days, I've been pretty sick. That's interesting the effects only work with certain integrators. I'll definitely look into that a bit more myself. I explicitly added IAS15 in the mass_loss example, and it still doesn't work. I'll try WHFAST too, but I don't think this will be the solution. I have a feeling it's going to be something really simple that I managed to overlook.

**Comment by Nofe4108** [comment]
In the mass_loss_sun_earth program I tried to switch the reboundx operator from mine to modify_mass, and it still didn't work, so something must be wrong with the way I'm adding the operator. Next I'm going to add a print statement to my mass_loss operator to see if the program is actually including it in the simulation like you suggested during our last meeting.

**Comment by Nofe4108** [comment]
Alright that worked. Also when I switched the mass loss effect back to the simple linear mass loss rate I had before it worked this time, and the Sun in the simulation did lose mass until it reached a certain value like I had intended it to. I'll update the repository on GitHub with all the code that worked. Now what do you think the next step should be? I know we want to try to match the simulation data we get with these added forces and effects to our previous python simulations over the summer.

**Comment by sabaronett** [comment]
@Nofe4108 , awesome! Glad it's working now.

By the way, for future comments, don't be afraid to directly reference commits like this https://github.com/sabaronett/reboundx/commit/9504f251e27b6d674381bc62f5fe1cceccbe9841 (simply copy+paste the commit URL), that way we can see exactly what each of us is talking about and its context. You can also "quote" lines of code from a file uploaded to our repo by clicking on the line number(s) (hold the shift-key for multiple lines), then the three dots and "copy permalink" (see image below from https://github.com/sabaronett/reboundx/blob/master/.gitignore).
<img src="https://user-images.githubusercontent.com/51060809/72865101-71778380-3c8b-11ea-8616-b34bbc7c6cd9.png" width="300px">
Paste the URL from the clipboard and you get this:
https://github.com/sabaronett/reboundx/blob/b74fa1c70f4ce5f0a83cbf8cc12ddb2b91e9e8d6/.gitignore#L1-L3

**Comment by sabaronett** [comment]
## `mass_loss` Tasks
@Nofe4108 , As for the next steps, tentatively:
1. Play around with add'l parameters (see https://reboundx.readthedocs.io/en/latest/add_effect.html#adding-parameters).
   1. Become really familiar with how to **register** add'l params in `src/core.c` (of different types, e.g., `int`, `double`), how to **access** them in `src/mass_loss.c`, and how to **set** them from `problem.c`.
   2. Once you got that down, in `problem.c` try writing a _super basic_ function outside of `int main()` that will return a value for the mass given some time `sim->t`. Then you can call that function within `heartbeat` by assigning its output to the add'l param used by your effect, e.g., `rebx_set_param_double(rebx, &sim->particles[1].ap, "mass", mass_function(sim->t));` This way you can test whether adjusting parameter values at each time step works. An example function could be:
```c
/* 
   Basic linear function (placed before int main()) 
   to return the mass in Msun as a function 
   of simulation time 
*/
double mass_function (double t) {
     double a = 1e-6; // slope
     return (1 - a*t);
}
```
2. Assuming those work, next start looking into how to create **splines** in C. Use the GNU Scientific Library (GSL, https://www.gnu.org/software/gsl/).
   1. Start off simple (e.g., splining just a couple of points) and make sure you got the ins-and-outs down like we did with the SciPy's CubicSpline function.
   2. Then replace the function you wrote back in 1,ii. with the spline instead, and see if that all works.
   3. Finally, try splining our MESA mass data. We'll need to figure out how best to "load" the .data file (essentially a .txt file) into `problem.c`, in order to spline it, as part of this step.
3. Ideally, we'd ultimately like to "attach" to the star (e.g., `particle[0]`) the spline as self-contained add'l parameter. That way, once set, the effect `mass_loss.c` takes care of the work, even without an explicit `heartbeat` function.
   1. To do this, we'll need to introduce a _custom type_ parameter (i.e., _not_ an `int` or `double`), in particular, a _pointer_ to the spline. See https://reboundx.readthedocs.io/en/latest/add_effect.html#adding-custom-types for more information.
   2. In the end, we may find this won't work. It will depend on what GSL's splines are, i.e., whether its a `struct`, or function, or something else. From my understanding, the add'l params must be some basic C data type or custom structure, otherwise this entire step might not work.

If we run into issues with step 3, we can ask Dr. Tamayo for advice and on what else to do. Let me know if you have any questions, get stuck, or run into any problems.

**Comment by Nofe4108** [comment]
Okay so I got a parameter to work in a simulation, so I next tried different ways to modify the parameter while the simulation was going. From what I've done so far it doesn't seem like you can change the parameter once the simulation starts, or if you can I just haven't found a way to do that yet. Let me know if you've found a way to do it. 

My idea now is that the parameters for the mass loss effect should just be a way for someone to input the file they want read and splined and a way to customize the spline that comes out. The spline will be made and attached to the simulation's star in the mass_loss.c code so there will be no need to use the additional parameter to attach the spline in the problem.c file. It won't be self-contained, but then there will be no need to worry about if making the spline a custom type parameter will even work.

Let me know if what I just said made sense, and if you think it'll work.

**Comment by sabaronett** [comment]
@Nofe4108 
Ah OK. No, I haven't yet tried modifying the parameters during a simulation yet, but I trust you.

Your idea makes sense, and it's definitely a faster way to implement the goals than trying to store the spline as a custom type parameter. One way to approach it is to pass as a string type parameter the path/filename (e.g., "`mass.txt`"), which you can then access from within `mass_loss.c`. Then you'll open, read, and spline the data all from within `mass_loss.c`.

The only issue with doing it this way is that it will _require_ GSL as a dependency to even compile REBOUNDx, which Dr. Tamayo may or may not want. I'm also not sure how well this complies with his "machine independence" [recommendation](https://reboundx.readthedocs.io/en/latest/add_effect.html#is-your-code-machine-independent). Perhaps he has some insight on this matter.

**Comment by dtamayo** [comment]
You definitely can change the parameter after the simulation starts, just call the set_param_double function (or one for other type) you used the first time and it will update it (you can check by calling get_param_double and seeing it changed).

pip manages python dependencies nicely, but adding C dependencies is a pain that most users will not have experience with, so I'd rather avoid it. I think the cleanest way to implement it is for your stellar evolution code to fit an interpolation function and store that to a file (mass_loss_history.bin) which your mass_loss.c function reads. That way you keep all the stellar evolution work you've done modular and separate from the N-body integration piece, and you don't have to worry about dependencies on the REBOUNDx side, I.e., you could send someone a mass_loss_history.bin file, and they could run it with their installation of REBOUNDx even without having any of the rest of your code.

One thing to think about is how smooth is your mass loss history? If it's smooth, then it's probably easiest to just store the coefficients of a fitting function (regular polynomials or maybe Chebyshev), so that mass_loss.c doesn't have to open and read from disk every timestep (will be much faster).

**Comment by sabaronett** [comment]
Thanks for that response @dtamayo. These sound like great ideas.

As for smoothness, below is an example of the Sun's mass loss data from MESA:
![Msun_vs_age](https://user-images.githubusercontent.com/51060809/73597899-1db83600-44e6-11ea-86de-62513ab6d929.png)

Back in the Python version, we found using SciPy's cubic spline interpolation methods to fit the data very well (without any under/over fitting artifacts), in addition to being quite fast to initially generate and retrieve values from at each timestep, although the spline was being held in memory.

I'll look into storing the interpolation function in a `.bin` file and reference it in our `[effects].c` as a start and get a sense of the performance from disk I/O ops at each timestep. From there I'll investigate storing the coefficients instead if feasible.

**Comment by dtamayo** [comment]
That seems like a very reasonable way to implement a python version. Porting to C will make it a factor of 2-3 faster, but you might lose part or all of that if you're opening and closing a .bin file every timestep. There are several ways to try and fix that, but it probably makes sense to get the easiest version to implement working first. Storing a .bin file from python might require some reverse engineering of their code to figure out the binary format (but maybe you can find someone that has already done that in C). Another possibility might be to take spline fitting code from Numerical Recipes in C.

As for smoothness, you just need it to be smooth over orbital timescales. I can't tell from the plot whether that's true...not sure how far apart the MESA outputs are or whether they're adaptively spaced with more outputs near the 'discontinuities'

**Comment by sabaronett** [comment]
@dtamayo 
The tidal drag effect I just finished successfully adding (fc3abd90b1926b29118451fc1a644dda7a32efd4) &mdash;albeit the parameters are hard-coded&mdash; is even more than 3 times faster. The simulation graphed in https://github.com/sabaronett/reboundx/issues/6#issuecomment-581066789 takes a little over 15 minutes to run in the Python version, but only took **3 seconds** in C!

I've opened #13 to track implementing `.bin` storage. So far I've come across Python's `pickle` [module ](https://docs.python.org/3/library/pickle.html), which can easily serialize/de-serialize, but it's intended only for Python, as far as "unpickling". There's also `dill` (which [extends ](https://pypi.org/project/dill/)`pickle`), which hints at some compatibility beyond Python, but I haven't confirmed that or tested it myself yet. That being said, I'm currently looking into C equivalents for pickling, as working in C will ultimately be needed for de-serialization into our `[effects.c]`. I'm guessing then we'll have to (re)implement some de-serialization source code into our `[effects.c]`, to avoid any new dependencies.

As for "smoothness," what would you consider to be an _orbital timescale_? I know for sure that MESA uses adaptive timesteps for their integrations. My concern with using polynomial interpolation is encountering Runge's phenomenon around those discontinuities (e.g., ~12.4 Gyr).

**Comment by dtamayo** [comment]
binaries are tricky in that you need to know the exact format in which they were written in order to read from them in C. I would think two easier ways to deal with this would be to

a) Figure out whether you can print out all the spline coefficients in Python, test that you can use those coefficients to manually. calculate the same values at arbitrary times that you get when using the default API (all in Python). Then store the coefficients in a text file that your REBOUNDx effect reads once the first time, and keeps in memory.

b) Write C code (e.g. from Numerical Recipes) that takes the MESA outputs and fits spline coefficients. Write a binary in C, so you then know how to read it back in.

a) may be easier!

Orbital timescale is just the orbital period of the innermost planet. If the mass it gets from MESA changes significantly within the timescale of one orbit, integrators like IAS15 will not converge.

**Comment by sabaronett** [comment]
Issue to be effectively resolved in #14  and #17.

Closing issue.
